src/Arcus.Security.Core/Providers/MutatedSecretNameCachedSecretProvider.cs

```diff
@@ -0,0 +1,139 @@
+﻿using System;
+using System.Threading.Tasks;
+using Arcus.Security.Core.Caching;
+using Arcus.Security.Core.Caching.Configuration;
+using GuardNet;
+using Microsoft.Extensions.Logging;
+
+namespace Arcus.Security.Core.Providers
+{
+    /// <summary>
+    /// Represents an <see cref="ICachedSecretProvider"/> that can mutate the secret name provided before looking up the secret.
+    /// </summary>
+    public class MutatedSecretNameCachedSecretProvider : MutatedSecretNameSecretProvider, ICachedSecretProvider
+    {
+        private readonly ICachedSecretProvider _implementation;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MutatedSecretNameSecretProvider"/> class.
+        /// </summary>
+        /// <param name="implementation">The actual <see cref="ISecretProvider"/> implementation to look up the secret.</param>
+        /// <param name="mutateSecretName">The function to mutate the name of the secret before looking up the secret.</param>
+        /// <param name="logger">The instance to log diagnostic messages during the secret name mutation.</param>
+        /// <exception cref="ArgumentNullException">
+        ///     Thrown when the <paramref name="implementation"/> or the <paramref name="mutateSecretName"/> is <c>null</c>.
+        /// </exception>
+        public MutatedSecretNameCachedSecretProvider(
+            ICachedSecretProvider implementation,
+            Func<string, string> mutateSecretName,
+            ILogger logger)
+            : base(implementation, mutateSecretName, logger)
+        {
+            Guard.NotNull(implementation, nameof(implementation), "Requires an secret provider instance to pass the mutated secret name to");
+            _implementation = implementation;
+        }
+
+        /// <summary>
+        /// Gets the cache-configuration for this instance.
+        /// </summary>
+        public ICacheConfiguration Configuration => _implementation.Configuration;
+
+        /// <summary>
+        /// Retrieves the secret value, based on the given name
+        /// </summary>
+        /// <param name="secretName">The name of the secret key</param>
+        /// <param name="ignoreCache">Indicates if the cache should be used or skipped</param>
+        /// <returns>Returns a <see cref="Task{TResult}"/> that contains the secret key</returns>
+        /// <exception cref="ArgumentException">The name must not be empty</exception>
+        /// <exception cref="ArgumentNullException">The name must not be null</exception>
+        /// <exception cref="SecretNotFoundException">The secret was not found, using the given name</exception>
+        public async Task<string> GetRawSecretAsync(string secretName, bool ignoreCache)
+        {
```

Comment:
Shouldn't this one call `GetSecretAsync` to reduce duplication?

Reply:
Hmm, no bc we want to call the `GetRawSecretAsync` on the inner `ISecretProvider` otherwise the `GetRawSecretAsync` on the inner implementation will never be called.

Reply:
Good point! Maybe we can change how we handle the task so that it's re-used and only call different operations on the secret provider here?

Reply:
Yes, exactly! Moved the common logic to a central method. Thanks. Resolving.